src/Contracts/Pools/OpdexStakingPool.cs

```diff
@@ -59,105 +54,137 @@ public UInt256 StakingRewardsBalance
     }
 
     /// <inheritdoc />
-    public UInt256 GetStakedBalance(Address address)
+    public UInt256 RewardPerStakedTokenLast
     {
-        return State.GetUInt256($"StakedBalance:{address}");
+        get => State.GetUInt256(nameof(RewardPerStakedTokenLast));
+        private set => State.SetUInt256(nameof(RewardPerStakedTokenLast), value);
     }
-
-    private void SetStakedBalance(Address address, UInt256 balance)
+    
+    /// <inheritdoc />
+    public UInt256 ApplicableStakingRewards
     {
-        State.SetUInt256($"StakedBalance:{address}", balance);
+        get => State.GetUInt256(nameof(ApplicableStakingRewards));
+        private set => State.SetUInt256(nameof(ApplicableStakingRewards), value);
     }
-
+    
     /// <inheritdoc />
-    public UInt256 GetStakedWeight(Address address)
+    public UInt256 GetStoredRewardPerStakedToken(Address address)
     {
-        return State.GetUInt256($"StakedWeight:{address}");
+        return State.GetUInt256($"RewardPerStakedToken:{address}");
     }
 
-    private void SetStakedWeight(Address address, UInt256 weight)
+    private void SeStoredRewardPerStakedToken(Address address, UInt256 reward)
     {
-        State.SetUInt256($"StakedWeight:{address}", weight);
+        State.SetUInt256($"RewardPerStakedToken:{address}", reward);
     }
-
+    
     /// <inheritdoc />
-    public UInt256 GetStakingRewards(Address staker)
+    public UInt256 GetStoredReward(Address address)
     {
-        return GetStakingRewardsExecute(staker, GetStakedBalance(staker));
+        return State.GetUInt256($"Reward:{address}");
+    }
+
+    private void SetStoredReward(Address address, UInt256 reward)
+    {
+        State.SetUInt256($"Reward:{address}", reward);
     }
     
     /// <inheritdoc />
-    public void Stake(UInt256 amount)
+    public UInt256 GetStakedBalance(Address address)
+    {
+        return State.GetUInt256($"StakedBalance:{address}");
+    }
+
+    private void SetStakedBalance(Address address, UInt256 balance)
+    {
+        State.SetUInt256($"StakedBalance:{address}", balance);
+    }
+
+    /// <inheritdoc />
+    public void StartStaking(UInt256 amount)
     {
         EnsureUnlocked();
         EnsureStakingEnabled();
-        
+        Assert(amount > 0, "OPDEX: CANNOT_STAKE_ZERO");
+
         MintStakingRewards(ReserveCrs, ReserveSrc);
         
-        var stakedBalance = GetStakedBalance(Message.Sender);
-        if (stakedBalance > 0)
-        {
-            CollectStakingRewardsExecute(Message.Sender, stakedBalance, false);
-            UnstakeExecute(Message.Sender, stakedBalance, false);
-        }
+        UpdateStakingPosition(Message.Sender);
+
+        TotalStaked += amount;
```

Comment:
```suggestion
        var updatedTotalStaked = TotalStaked += amount;
```

Small gas saving logging on line 120